File: day3/part2.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <string>
using namespace std;

int binary_to_decimal(string s) {
	int num = 0;
	for (char c : s) {
		num <<= 1;
		if (c == '1') {
			num += 1;
		}
	}
	return num;
}

int main() {
	freopen("input.txt", "r", stdin);
	vector<string> binary_strings;
	while (cin) {
		string s;
		cin >> s;
		binary_strings.push_back(s);
	}
	int num_bits = binary_strings[0].length();

	// Calculate oxygen generator rating
	unordered_set<string> possible_numbers;
	for (string s : binary_strings) {
		possible_numbers.insert(s);
	}
	vector<int> current_indices;
	for (int i = 0; i < binary_strings.size(); i++) {
		current_indices.push_back(i);
	}
	for (int i = 0; i < num_bits; i++) {
		// Calculate most frequent bit
		int frequency = 0;
		for (int idx : current_indices) {
			if (binary_strings[idx][i] == '1') {
				frequency++;
			}
		}
		char bit;
		if (frequency*2 >= current_indices.size()) {
			bit = '1';
		} else {
			bit = '0';
		}
		// Remove non-matching binary strings
		vector<int> next_indices;
		for (int idx : current_indices) {
			if (binary_strings[idx][i] == bit) {
				next_indices.push_back(idx);
			} else {
				possible_numbers.erase(binary_strings[idx]);
				if (possible_numbers.size() == 1) {
					next_indices = {};
					break;
				}
			}
		}
		current_indices = next_indices;
	}
	int oxygen_rating = binary_to_decimal(*(possible_numbers.begin()));

	// Calculate CO2 scrubber rating
	possible_numbers = {};
	for (string s : binary_strings) {
		possible_numbers.insert(s);
	}
	current_indices = {};
	for (int i = 0; i < binary_strings.size(); i++) {
		current_indices.push_back(i);
	}
	for (int i = 0; i < num_bits; i++) {
		// Calculate most frequent bit
		int frequency = 0;
		for (int idx : current_indices) {
			if (binary_strings[idx][i] == '1') {
				frequency++;
			}
		}
		char bit;
		if (frequency*2 >= current_indices.size()) {
			bit = '0';
		} else {
			bit = '1';
		}
		// Remove non-matching binary strings
		vector<int> next_indices;
		for (int idx : current_indices) {
			if (binary_strings[idx][i] == bit) {
				next_indices.push_back(idx);
			} else {
				possible_numbers.erase(binary_strings[idx]);
				if (possible_numbers.size() == 1) {
					next_indices = {};
					break;
				}
			}
		}
		current_indices = next_indices;
	}
	int scrubber_rating = binary_to_decimal(*(possible_numbers.begin()));

	int life_support_rating = oxygen_rating * scrubber_rating;
	cout << "Result: " << life_support_rating << '\n';
}

File: day3/part2.py
def binary_to_decimal(s):
    num = 0
    for c in s:
        num <<= 1
        if c == '1':
            num += 1
    return num

# Read file input
with open("input.txt", "r") as f:
    data = [str(i).rstrip('\n') for i in f]
num_bits = len(data[0])

# Calculate oxygen generator rating
possible_numbers = set(data)
current_indices = [i for i in range(len(data))]
for i in range(num_bits):
    # Calculate most frequent bit
    frequency = 0
    for idx in current_indices:
        if data[idx][i] == '1':
            frequency += 1
    if frequency >= len(current_indices)/2:
        bit = '1'
    else:
        bit = '0'
    # Remove non-matching binary strings
    next_indices = []
    for idx in current_indices:
        if data[idx][i] == bit:
            next_indices.append(idx)
        else:
            possible_numbers.remove(data[idx])
            if len(possible_numbers) == 1:
                next_indices = []
                break
    current_indices = next_indices
oxygen_rating = binary_to_decimal(list(possible_numbers)[0])

# Calculate CO2 scrubber rating
possible_numbers = set(data)
current_indices = [i for i in range(len(data))]
for i in range(num_bits):
    # Calculate least frequent bit
    frequency = 0
    for idx in current_indices:
        if data[idx][i] == '1':
            frequency += 1
    if frequency >= len(current_indices)/2:
        bit = '0'
    else:
        bit = '1'
    # Remove non-matching binary strings
    next_indices = []
    for idx in current_indices:
        if data[idx][i] == bit:
            next_indices.append(idx)
        else:
            possible_numbers.remove(data[idx])
            if len(possible_numbers) == 1:
                next_indices = []
                break
    current_indices = next_indices
scrubber_rating = binary_to_decimal(list(possible_numbers)[0])

life_support_rating = oxygen_rating * scrubber_rating
print(f"Result: {life_support_rating}")
